lib/activerecord-tenant-level-security/schema_dumper.rb

```diff
@@ -28,23 +28,35 @@ def policies_in_database
           tablename;
       SQL
       results = ActiveRecord::Base.connection.execute(query)
-      table_names = results.map { |x| x["tablename"] }
 
-      table_names.map { |t| Policy.new(t) }
+      results.map { |x| Policy.new(x["tablename"], x["qual"]) }
     end
 
     class Policy
-      def initialize(table_name)
+      def initialize(table_name, qual)
         @table_name = table_name
+        @qual = qual
       end
 
       def to_schema
-        %(  create_policy "#{table_name}")
+        %(  create_policy "#{table_name}", tenant_id: :#{extract_type_from_qual})
       end
 
       private
 
       attr_reader :table_name
+
+      def extract_type_from_qual
+        # qual is a string like:  (tenant_id = (NULLIF(current_setting('tenant_level_security.tenant_id'::text), ''::text))::integer)
+        # We want to extract the type, which is the last part of the string
+        # We just extract specific sequence of characters, because we know the format of the string
+        if @qual[0..90] == "(tenant_id = (NULLIF(current_setting('tenant_level_security.tenant_id'::text), ''::text))::" &&
+            @qual[-1] == ")"
+          @qual[91..-2]
+        else
+          raise "Unexpected qual: #{@qual}"
+        end
+      end
```

Comment:
This looks so complicated...

Looking at Active Record's implementation, it seems that it doesn't dump schemas from such complex SQL, and only treats simple representations as migration.
https://github.com/rails/rails/blob/v7.0.4.3/activerecord/lib/active_record/schema_dumper.rb

Perhaps instead of adding the `tenant_id` to the `create_policy`, there might be a way to internally check the `tenant_id` column type and switch the policy condition. What do you think?

Reply:
Thanks.
I fixed this.
Now, we automatically detect data type of `tenant_id` using `#columns` method here.
https://github.com/kufu/activerecord-tenant-level-security/blob/cd39feb362498bc5afae11df6de638b82405f03a/lib/activerecord-tenant-level-security/schema_statements.rb#L30-L35